process/slash/detector/multipleHeaderProposalsDetector.go

```diff
@@ -0,0 +1,239 @@
+	cache := newRoundProposerDataCache(CacheSize)
+	return &multipleHeaderProposalsDetector{
+		cache:            cache,
+		nodesCoordinator: nodesCoordinator,
+		roundHandler:     roundHandler,
+	}, nil
+}
+
+// VerifyData - checks if an intercepted data(which should be a header) represents a slashable event.
+// If another header with the same round and proposer exists, but a different hash, then a proof of type
+// slash.MultipleProposal is provided, otherwise a nil proof, along with an error is provided indicating that
+// no slashing event has been detected or an error occurred verifying the data.
+func (mhp *multipleHeaderProposalsDetector) VerifyData(data process.InterceptedData) (slash.SlashingProofHandler, error) {
+	header, castOk := data.(*interceptedBlocks.InterceptedHeader)
+	if !castOk {
+		return nil, process.ErrCannotCastInterceptedDataToHeader
+	}
+
+	round := header.HeaderHandler().GetRound()
+	if !mhp.isRoundRelevant(round) {
+		return nil, process.ErrHeaderRoundNotRelevant
+	}
+
+	proposer, err := mhp.getProposerPubKey(header.HeaderHandler())
+	if err != nil {
+		return nil, err
+	}
+
+	slashType, slashLevel, headers := mhp.getSlashingResult(header, round, proposer)
+	mhp.cache.add(round, proposer, header)
+
+	if slashType == slash.MultipleProposal {
+		return slash.NewMultipleProposalProof(slashType, slashLevel, headers)
+	}
+	return nil, process.ErrNoSlashingEventDetected
+}
+
+func (mhp *multipleHeaderProposalsDetector) isRoundRelevant(headerRound uint64) bool {
+	currRound := uint64(mhp.roundHandler.Index())
+	return absDiff(currRound, headerRound) < MaxDeltaToCurrentRound
+}
+
+func absDiff(x, y uint64) uint64 {
+	if x < y {
+		return y - x
+	}
+	return x - y
+}
+
+func (mhp *multipleHeaderProposalsDetector) getProposerPubKey(header data.HeaderHandler) ([]byte, error) {
+	validators, err := mhp.nodesCoordinator.ComputeConsensusGroup(
+		header.GetRandSeed(),
+		header.GetRound(),
+		header.GetShardID(),
+		header.GetEpoch())
+
+	if err != nil {
+		return nil, err
+	}
+	if len(validators) == 0 {
+		return nil, process.ErrEmptyConsensusGroup
+	}
+
+	return validators[0].PubKey(), nil
+}
+
+func (mhp *multipleHeaderProposalsDetector) getSlashingResult(
+	currHeader process.InterceptedData,
+	currRound uint64,
+	proposerPubKey []byte,
+) (slash.SlashingType, slash.SlashingLevel, []process.InterceptedData) {
+	headers := make([]process.InterceptedData, 0)
+	slashType := slash.None
+	slashLevel := slash.Level0
+	proposedHeaders := mhp.cache.proposedData(currRound, proposerPubKey)
+
+	if len(proposedHeaders) >= 1 {
+		headers = mhp.getProposedHeadersWithDifferentHash(currHeader.Hash(), proposedHeaders)
+		if len(headers) >= 1 {
+			// TODO: Maybe a linear interpolation to deduce severity?
+			if len(headers) == 1 {
+				slashLevel = slash.Level1
+			} else {
+				slashLevel = slash.Level2
+			}
+			slashType = slash.MultipleProposal
+			headers = append(headers, currHeader)
```

Comment:
I think here there is the possibility to append the same header multiple times, if you already have len(proposedHeaders) >= 1

then for example you might incorrectly increase the slashLevel to 2 even if the current header is already in the list of proposedHeaders.